### PR TITLE
extended job APIs

### PR DESCRIPTION
**Pull request by blublinsky**
## Why are these changes needed?

In the Job APIs for the apiserver added job submitter parameters. This fixes the problem of Python/Ray version mismatch in the case of existing Ray cluster

## Checks

- [ x] I've made sure the tests are passing. 
- Testing Strategy
   - [ x] Unit tests
   - [x] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by blublinsky** [comment]
It requires rebasing

**Comment by blublinsky** [comment]
@z103cb, @astefanutti, pls take a look

cc: @kevin85421

**Review by z103cb — COMMENTED** [review]
There are two things, that I think should be included in this PR:
* update of the README.md with examples
* the [func ValidateCreateJobRequest(request *api.CreateRayJobRequest) error {](https://github.com/z103cb/kuberay/blob/a8895aeb6801e2943a0940605625c51094949698/apiserver/pkg/server/job_server.go#L110) function should be enhanced to validate the required fields in the ray submitter is present in the request. I would go as far as making the object required if a cluster selector is used.

File: apiserver/pkg/util/job.go [hunk 1]
```diff
@@ -39,6 +41,51 @@ func NewRayJob(apiJob *api.RayJob, computeTemplateMap map[string]*api.ComputeTem
 		}
 		rayJob.Spec.RayClusterSpec = clusterSpec
 	}
+	if apiJob.JobSubmitter != nil {
+		// Job submitter is specified, create
```
[the remark this answers is not in the document]
Comment (blublinsky) [reply apiserver/pkg/util/job.go:45]: fixed

**Comment by blublinsky** [comment]
> * the [func ValidateCreateJobRequest(request *api.CreateRayJobRequest) error {](https://github.com/z103cb/kuberay/blob/a8895aeb6801e2943a0940605625c51094949698/apiserver/pkg/server/job_server.go#L110) function should be enhanced to validate the required fields in the ray submitter is present in the request. I would go as far as making the object required if a cluster selector is used.

Implemented check in util/job.go

**Comment by blublinsky** [comment]
Add documentation on how to use it in Readme

File: apiserver/README.md [hunk 2]
```diff
@@ -1084,6 +1084,244 @@ Examples:
...
+The above example creates a new Ray cluster, executes a job on it and optionally deletes a cluster. As an alternative, the same command allows creating a new job on the existing cluster by referencing it in the payload.
+
+Examples:
+
+Start from creating Ray cluster (We assume here that the [template](test/cluster/template/simple) and [configmap](test/job/code.yaml) are already created).
+
+* Request
+  
+```sh
+curl -X POST 'localhost:8888/apis/v1alpha2/namespaces/default/clusters' \
```
Comment (z103cb) [inline apiserver/README.md:1096]: can you change the port to 31888 so that all the curl commands are consistent in this document?

**Review by z103cb — APPROVED** [review]
LGTM

File: apiserver/README.md (hunk 2 above)
Comment (blublinsky) [reply apiserver/README.md]: Done

**Comment by blublinsky** [comment]
@astefanutti all ready for you

File: apiserver/pkg/model/converter.go [hunk 3]
```diff
@@ -374,6 +374,27 @@ func FromCrdToApiJob(job *rayv1api.RayJob) (pbJob *api.RayJob) {
...
+	if job.Spec.SubmitterPodTemplate != nil {
+		pbJob.JobSubmitter = &api.RayJobSubmitter{
+			Image: job.Spec.SubmitterPodTemplate.Spec.Containers[0].Image,
+		}
+		if job.Spec.SubmitterPodTemplate.Spec.Containers[0].Resources.Limits.Cpu().String() != "1" {
+			pbJob.JobSubmitter.Cpu = job.Spec.SubmitterPodTemplate.Spec.Containers[0].Resources.Limits.Cpu().String()
+		}
+		if job.Spec.SubmitterPodTemplate.Spec.Containers[0].Resources.Limits.Memory().String() != "1Gi" {
+			pbJob.JobSubmitter.Memory = job.Spec.SubmitterPodTemplate.Spec.Containers[0].Resources.Limits.Memory().String()
+		}
```
Comment (astefanutti) [inline apiserver/pkg/model/converter.go:386]:
```suggestion
		if cpu := job.Spec.SubmitterPodTemplate.Spec.Containers[0].Resources.Limits.Cpu().String(); cpu != "1" {
			pbJob.JobSubmitter.Cpu = cpu
		}
		if memory := job.Spec.SubmitterPodTemplate.Spec.Containers[0].Resources.Limits.Memory().String(); memory != "1Gi" {
			pbJob.JobSubmitter.Memory = memory
		}
```

File: apiserver/pkg/util/job.go [hunk 4]
```diff
@@ -39,6 +47,51 @@ func NewRayJob(apiJob *api.RayJob, computeTemplateMap map[string]*api.ComputeTem
...
+			cpus = apiJob.JobSubmitter.Cpu
+		}
+		if apiJob.JobSubmitter.Memory != "" {
+			memorys = apiJob.JobSubmitter.Memory
+		}
+		rayJob.Spec.SubmitterPodTemplate = &v1.PodTemplateSpec{
+			Spec: v1.PodSpec{
+				Containers: []v1.Container{
+					{
+						Name:  apiJob.Name + "-submitter",
```
Comment (astefanutti) [inline apiserver/pkg/util/job.go:64]: What's the need to make the container name dynamic? Generally it's preferable to keep them static.

**Review by astefanutti — COMMENTED** [review]
Left a couple of comments. Will LGTM after that.

File: apiserver/pkg/util/job.go (hunk 4 above)
Comment (blublinsky) [reply apiserver/pkg/util/job.go:64]: To avoid name collision in case we want to run multiple jobs against the same cluster

Comment (astefanutti) [reply apiserver/pkg/util/job.go:64]: How does the container name related to the ray job submission id?

File: apiserver/pkg/model/converter.go (hunk 3 above)
Comment (blublinsky) [reply apiserver/pkg/model/converter.go]: Fixed.

**Review by kevin85421 — APPROVED** [review]

**Pull request merged**